filebeat/beat/spooler.go

```diff
@@ -9,115 +9,134 @@ import (
 	"github.com/elastic/beats/libbeat/logp"
 )
 
-type Spooler struct {
-	Filebeat           *Filebeat
-	exit               chan struct{}
-	nextFlushTime      time.Time
-	nextFlushTimeMutex *sync.Mutex
-	spool              []*input.FileEvent
-	Channel            chan *input.FileEvent
-}
+var debugf = logp.MakeDebug("spooler")
 
-func NewSpooler(filebeat *Filebeat) *Spooler {
-	spooler := &Spooler{
-		Filebeat:           filebeat,
-		exit:               make(chan struct{}),
-		nextFlushTimeMutex: &sync.Mutex{},
-	}
+// channelSize is the number of events Channel can buffer before blocking will occur.
+const channelSize = 16
 
-	// Set the next flush time
-	spooler.setNextFlushTime()
-	spooler.Channel = make(chan *input.FileEvent, 16)
+// Spooler aggregates the events and sends the aggregated data to the publisher.
+type Spooler struct {
+	Channel chan *input.FileEvent // Channel is the input to the Spooler.
 
-	return spooler
-}
+	// Config
+	idleTimeout time.Duration // How often to flush the spooler if spoolSize is not reached.
+	spoolSize   uint64        // Maximum number of events that are stored before a flush occurs.
 
-func (spooler *Spooler) Config() error {
-	config := &spooler.Filebeat.FbConfig.Filebeat
+	exit          chan struct{}             // Channel used to signal shutdown.
+	nextFlushTime time.Time                 // Scheduled time of the next flush.
+	publisher     chan<- []*input.FileEvent // Channel used to publish events.
+	spool         []*input.FileEvent        // FileEvents being held by the Spooler.
+	wg            sync.WaitGroup            // WaitGroup used to control the shutdown.
+}
 
-	// Set default pool size if value not set
-	if config.SpoolSize == 0 {
-		config.SpoolSize = cfg.DefaultSpoolSize
+// NewSpooler creates and returns a new Spooler. The returned Spooler must be
+// started by calling Start before it can be used.
+func NewSpooler(
+	config cfg.FilebeatConfig,
+	publisher chan<- []*input.FileEvent,
+) *Spooler {
+	spoolSize := config.SpoolSize
+	if spoolSize <= 0 {
+		spoolSize = cfg.DefaultSpoolSize
+		debugf("Spooler will use the default spool_size of %d", spoolSize)
 	}
 
-	// Set default idle timeout if not set
-	if config.IdleTimeout == "" {
-		logp.Debug("spooler", "Set idleTimeoutDuration to %s", cfg.DefaultIdleTimeout)
-		// Set it to default
-		config.IdleTimeoutDuration = cfg.DefaultIdleTimeout
-	} else {
-		var err error
-
-		config.IdleTimeoutDuration, err = time.ParseDuration(config.IdleTimeout)
-
-		if err != nil {
-			logp.Warn("Failed to parse idle timeout duration '%s'. Error was: %v", config.IdleTimeout, err)
-			return err
-		}
+	idleTimeout := config.IdleTimeout
+	if idleTimeout <= 0 {
```

Comment:
Invalid `time.Duratation's` like `5 sec` or `5 seconds` are caught during the YAML unmarshaling and result in an error. As currently written here, valid durations like `-5s` are treated as if the value wasn't set.

I think there should be some additional bounds checking on these values like verifying that the idle timeout is greater than `x ms`. I just wasn't sure what that minimum value should be or if that is behavior you want? If we do add this check, should the code fail, or should it just set the value to the minimum allowable value?